composition-js/src/merging/merge.ts

```diff
@@ -1102,40 +1106,55 @@ class Merger {
   }
 
   private addArgumentsShallow<T extends FieldDefinition<any> | DirectiveDefinition>(sources: (T | undefined)[], dest: T) {
+    const argNames = new Set<string>();
     for (const source of sources) {
       if (!source) {
         continue;
       }
-      for (const argument of source.arguments()) {
-        if (!dest.argument(argument.name)) {
-          dest.addArgument(argument.name);
-        }
-      }
-    }
-  }
-
-  private mergeArgument(sources: (ArgumentDefinition<any> | undefined)[], dest: ArgumentDefinition<any>, useIntersection: boolean = false) {
-    if (useIntersection) {
-      for (const source of sources) {
-        if (!source) {
+      source.arguments().forEach((arg) => argNames.add(arg.name));
+    }
+
+    for (const argName of argNames) {
+      // We add the arguement unconditionally even if we're going to remove it in
+      // some path. Done because this helps reusing our "reportMismatchHint" method
+      // in those cases.
+      const arg = dest.addArgument(argName);
+      // If all the sources that have the field have the argument, we do merge it
+      // and we're good, but otherwise ...
+      if (!sources.every((s) => !s || s.argument(argName))) {
```

Comment:
**nitpicky question** Could this be changed to

```suggestion
      if (sources.some((s) => !s || s.argument(argName))) {
```

I think that's the same logic, but it's a little easier for my brain to understand that version.

Reply:
Right. I initially had the `if` without the negation and then a `else`, but realised I didn't need the first part. Agreed avoiding the negation is probably better now. That said, to have the same logic I believe it needs to be:
```
if (sources.some((s) => s && !s.argument(argName))) {
```
Essentially, `Not(For all x, P(x))` is `It exists x, Not(P(x))`.

Reply:
Right, good catch on the logic. Makes sense to me if you wanna make that tweak.